core/metrics/src/main/scala/org/typelevel/otel4s/metrics/BatchCallback.scala

```diff
@@ -0,0 +1,305 @@
+    * multiple asynchronous instruments.
+    *
+    * The callback will be called when the instruments are being observed.
+    *
+    * Callbacks are expected to abide by the following restrictions:
+    *   - Short-living and (ideally) non-blocking
+    *   - Run in a finite amount of time
+    *   - Safe to call repeatedly, across multiple threads
+    *
+    * @example
+    *   {{{
+    * val meter: Meter[F] = ???
+    * val server: F[Unit] = ??? // runs the server
+    *
+    * val background: Resource[F, Unit] =
+    *   for {
+    *     counter <- Resource.eval(meter.observableCounter[Long]("counter").createObserver)
+    *     upDownCounter <- Resource.eval(meter.observableUpDownCounter[Double]("up-down-counter").createObserver)
+    *     gauge <- Resource.eval(meter.observableGauge[Double]("gauge").createObserver)
+    *     callback = counter.record(1L) *> upDownCounter.record(2.0) *> gauge.record(3.0)
+    *     _ <- meter.batchCallback(callback, counter, upDownCounter, gauge)
+    *   } yield ()
+    *
+    * background.surround(server) // register batch callback and run the server
+    *   }}}
+    *
+    * @param callback
+    *   the callback to to observe values on-demand
+    *
+    * @param observable
+    *   the instrument for which the callback may observe values
+    *
+    * @param rest
+    *   the instruments for which the callback may observe values
+    */
+  def apply(
+      callback: F[Unit],
+      observable: ObservableMeasurement[F, _],
+      rest: ObservableMeasurement[F, _]*
+  ): Resource[F, Unit]
+
+  /** Constructs a batch callback.
+    *
+    * Batch callbacks allow a single callback to observe measurements for
+    * multiple asynchronous instruments.
+    *
+    * The callback will be called when the instruments are being observed.
+    *
+    * Callbacks are expected to abide by the following restrictions:
+    *   - Short-living and (ideally) non-blocking
+    *   - Run in a finite amount of time
+    *   - Safe to call repeatedly, across multiple threads
+    *
+    * @example
+    *   {{{
+    * val meter: Meter[F] = ???
+    * val server: F[Unit] = ??? // runs the server
+    *
+    * val background: Resource[F, Unit] =
+    *   meter.batchCallback.of(
+    *     meter.observableCounter[Long]("counter").createObserver,
+    *     meter.observableUpDownCounter[Double]("up-down-counter").createObserver,
+    *     meter.observableGauge[Double]("gauge").createObserver
+    *   ) { (counter, upDownCounter, gauge) =>
+    *     counter.record(1L) *> upDownCounter.record(2.0) *> gauge.record(3.0)
+    *   }
+    *
+    * background.surround(server) // register batch callback and run the server
+    *   }}}
+    */
+  final def of[A1, A2](
+      a1: F[ObservableMeasurement[F, A1]],
+      a2: F[ObservableMeasurement[F, A2]]
+  )(
+      cb: (
+          ObservableMeasurement[F, A1],
+          ObservableMeasurement[F, A2]
+      ) => F[Unit]
+  )(implicit F: Apply[F]): Resource[F, Unit] =
```

Comment:
Technically, we don't need to require `F: Apply[F]` here. But then we need to copy-paste the `of` methods in 3 places: noop instance, oteljava implementation, SDK implementation (upcoming).